Script: Fields API for Dense Vector

stu-elastic opened the pull request:
Adds the fields API for `dense_vector` field mapper.

Adds a `DenseVector` interface for the value type.

Implemented by:
 * `KnnDenseVector` which wraps a decoded float array from `VectorValues`
 * `BinaryDenseVector` which lazily decodes a `BytesRef` from `BinaryDocValues`

The vector operations have moved into those implements from `BinaryDenseVectorScriptDocValues.java` and  `KnnDenseVectorScriptDocValues.java`, respectively.

The `DenseVector` API is:
```
float getMagnitude();
double dotProduct(float[] | List);
double l1Norm(float[] | List);
double l2Norm(float[] | List);
float[] getVector();
int dims();

boolean isEmpty(); // does the value exist
int size();        // 0 if isEmpty(), 1 otherwise
Iterator<Float> iterator()
```

`dotProduct`, `l1Norm` and `l2Norm` take a `float[]` or a `List` via the
a delegating `default` method on the `DenseVector` interface.

The `DenseVectorDocValuesField` abstract class contains two getter APIS.
It is implemented by  `KnnDenseVectorDocValuesField` and
`BinaryDenseVectorDocValuesField`.

```
DenseVector get()
DenseVector get(DenseVector defaultValue)
```

The `get()` method is included because there isn't a good default dense vector,
so that API returns an empty `DenseVector` which throws an
`IllegalArgumentException` for all method calls other than `isEmpty()`,
`size()` and `iterator()`.

The empty dense vector will always be `DenseVector.EMPTY` in case users want
to use equality checks.

Refs: #79105

stu-elastic committed:
Script: Fields API for Dense Vector

Adds the fields API for `dense_vector` field mapper.

Adds a `DenseVector` interface for the value type.

Implemented by:
 * `KnnDenseVector` which wraps a decoded float array from `VectorValues`
 * `BinaryDenseVector` which lazily decodes a `BytesRef` from `BinaryDocValues`

The `DenseVector` API is simliar to `BinaryDenseVectorScriptDocValues`.

```
float getMagnitude();
double dotProduct(float[]);
double l1Norm(float[]);
double l2Norm(float[]);
float[] getVector();
int dims();

boolean isEmpty(); // does the value exist
int size();        // 0 if isEmpty(), 1 otherwise
PrimitiveIterator.OfDouble iterator()
```

The `DenseVectorDocValuesField` abstract class contains two getter APIS.
It is implemented by  `KnnDenseVectorDocValuesField` and
`BinaryDenseVectorDocValuesField`.

```
DenseVector get()
DenseVector get(DenseVector)
```

The `get()` method is included because there isn't a good default dense vector,
so that API returns an empty `DenseVector` which throws an
`IllegalArgumentException` for all method calls other than `isEmpty()`,
`size()` and `iterator()`.

The empty dense vector will always be `DenseVector.EMPTY` in case users want
to use equality checks.

Refs: #79105

stu-elastic committed:
Rearrange args, remove comment

elasticsearchmachine commented:
Hi @stu-elastic, I've created a changelog YAML for you.

stu-elastic committed:
Update docs/changelog/83550.yaml

stu-elastic committed:
Unit tests, yaml tests, take List<Number> or float[]

stu-elastic committed:
Merge branch 'pain_field_dense_vector' of github.com:stu-elastic/elasticsearch into pain_field_dense_vector

stu-elastic committed:
Use QueryVector wrapper

stu-elastic committed:
Merge into same yaml

stu-elastic committed:
spotless

stu-elastic committed:
yaml tests, merge query vector into one file

stu-elastic committed:
yaml test for iterator over missing vectors

stu-elastic committed:
Handle null VectorValues for empty in KnnDenseVectorDocValuesField

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into pain_field_dense_vector

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

stu-elastic commented:
@jtibshirani can you particularly take a look to make sure I moved the vector operations `dotProduct`, `l1Norm`, etc over correctly? `KnnDenseVectorScriptDocValues` -> `KnnDenseVector`, `BinaryDenseVectorScriptDocValues` -> `BinaryDenseVector`

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVector.java:
```diff
@@ -0,0 +1,113 @@
+    float getMagnitude();
+
+    // The implementations of these vector operations
+    double dotProduct(QueryVector queryVector);
+
+    double l1Norm(QueryVector queryVector);
+
+    double l2Norm(QueryVector queryVector);
```
Should any of these take an additional boolean parameter to normalize the QueryVector? Or should we document that these expect the QueryVector to be normalized?

stu-elastic replied:
Those don't need to be normalized, is my understanding.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVector.java:
```diff
@@ -0,0 +1,113 @@
+public interface DenseVector extends Iterable<Float> {
+    float getMagnitude();
+
+    // The implementations of these vector operations
```
This is not a helpful comment.

stu-elastic replied:
Removed.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVectorDocValuesField.java:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Get the DenseVector for a document if one exists, DenseVector.EMPTY otherwise
+     */
+    public abstract DenseVector get();
```
Are we should we want to add this now? It may be beneficial to figure out how we plan to implement implicit default values for the scripting fields API in general first since deprecation is a long-term process if we decide to do something different.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVectorScriptDocValues.java:
```diff
@@ -10,85 +10,75 @@
         return dims;
     }
 
+    private DenseVector getVectorChecked() {
```
Maybe rename `getCheckedVector`?

stu-elastic replied:
Renamed.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/QueryVector.java:
```diff
@@ -0,0 +1,95 @@
+            throw new IllegalArgumentException(badElement(element, i));
+        }
+
+        protected static String badElement(Object element, int index) {
```
Looks like this is only used a single time so please just move the message to the `get` method.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/QueryVector.java:
```diff
@@ -0,0 +1,95 @@
+        return new ListQueryVector(vector);
+    }
+
+    public static QueryVector fromObject(Object vector) {
```
May want a comment that is also for Painless support as we do not have type overloading.

stu-elastic replied:
Added.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/VectorDVLeafFieldData.java:
```diff
@@ -64,30 +58,14 @@ public SortedBinaryDocValues getBytesValues() {
             if (indexed) {
                 VectorValues values = reader.getVectorValues(field);
                 if (values == null || values == VectorValues.EMPTY) {
```
You don't need the null check here anymore.

jdconrad commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/VectorDVLeafFieldData.java:
```diff
@@ -64,30 +58,14 @@ public SortedBinaryDocValues getBytesValues() {
-                            return 0;
-                        }
-                    }, dims), name);
+                    // There's no way for KnnDenseVectorDocValuesField to reliably differentiate between VectorValues.EMPTY and
+                    // values that can be iterated through. Since VectorValues.EMPTY throws on docID(), pass a null instead.
+                    values = null;
```
I find this behavior to be a bit strange in that `VectorValues.EMPTY` will throw an exception, but `null` does not.

stu-elastic replied:
`VectorValues.EMPTY` is part of the Lucene API, so it cannot be easily changed as part of this PR

jdconrad commented on x-pack/plugin/vectors/src/main/resources/org/elasticsearch/xpack/vectors/query/org.elasticsearch.xpack.vectors.txt:
```diff
@@ -11,6 +11,40 @@ class org.elasticsearch.xpack.vectors.query.DenseVectorScriptDocValues {
+    DenseVector get(DenseVector)
+}
+
+# implementation of DenseVectorDocValuesField
+class org.elasticsearch.xpack.vectors.query.KnnDenseVectorDocValuesField {
+}
+
+# implementation of DenseVectorDocValuesField
+class org.elasticsearch.xpack.vectors.query.BinaryDenseVectorDocValuesField {
+}
+
```
While not necessary as these are top-level field classes, for consistency we have duck-typed all of them so far.

jdconrad reviewed (approved):
This looks good to me from the scripting perspective. Left some comments. Please wait for @jtibshirani to review before committing.

stu-elastic commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVector.java:
```diff
@@ -0,0 +1,113 @@
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+public interface DenseVector extends Iterable<Float> {
```
I ported over the `ScriptDocValues` API which was also missing this.  We'd have to normalize the query vector.  @jtibshirani, what are your thoughts?

stu-elastic replied:
Done.

stu-elastic committed:
rm dv comment, getCheckedVector, qv comment, rm badElement helper, rm vv null check

jtibshirani reviewed (commented):
Thanks @stu-elastic, this looks good to me overall. I had one high-level question about the `QueryVector` class -- why do we need this class? Can we just force the user to pass in a float array? Or always convert to an array up-front? In our testing we found that converting to a float array upfront was good for performance, since you avoid the unboxing costs of iterating over a `List<Float>` .

stu-elastic commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVectorDocValuesField.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    public abstract DenseVector get();
+
+    public abstract DenseVector get(DenseVector defaultValue);
```
We'd like it to be consistent.  There'd only be two "defaults" available right now, `DenseVector.EMPTY` and `null`, which is unfortunate but the nature of this type.

stu-elastic commented:
> I don't think it's helpful to extend Iterable<Float>, since you can just retrieve the vector through getVector.

The [Field interface](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/script/field/Field.java#L12) extends `Iterable`, because most fields are multi-valued.  Since it doesn't make sense in this case, I'll have it throw `UnsupportedOperationException` and we'll make iterable duck typed later on.

> I liked the old name vectorValue a little more. Currently you'd write DenseVector#getVector which seems kind of weird since the object is already a vector?

Users rarely see the class name in painless, check out the [rest tests](https://github.com/elastic/elasticsearch/pull/83550/files#diff-22467e3575bece7728016800dc5d82b238ce3c6797a196b37806df5764b9c4d8R228) to see how that ends up:
```field('dv').get().vector[2]```, that's why I changed the name to vector.  I agree it looks weird in Java.

I can change it to `field('dv').get().vectorValue[2]` if you'd prefer.

> I agree we should include cosineSimilarity, this was available in the old API and is something people will expect.
> Only cosine similarity needs to normalize the vector (and this should happen "behind the scenes").

cosine similarity is:
`dotProduct(queryVector) / denseVector.getMagnitude()`

The query vector will be normalized on every API call in this case.  I can do the normalization without allocating a new `float[]` for the query vector for [BinaryDenseVector](https://github.com/elastic/elasticsearch/pull/83550/files#diff-5b31cd7b08eb0ba538574d4c6ea439cdc3c9564069c0648a6ea886f1a6a9c959R41).  However, [KnnDenseVector](https://github.com/elastic/elasticsearch/pull/83550/files#diff-631c0702d086c91cd6c571a753082f0b99b9899304b8a1662444a04382824b82R34) calls into Lucene's `VectorUtil.cosine` so I'll have to create a new `float[]` for the normalized query vector.

[`ScoreScriptUtils.cosineSimilarity()` ](https://github.com/elastic/elasticsearch/pull/83550/files#diff-3d28338bcfb4a99c333f39b8f9e15efa817c35cd969f69490cb73694115f4e76R126) will still work and will avoid this penalty for `KnnDenseVector`.

stu-elastic commented:
Re: `QueryVector`

> why do we need this class? Can we just force the user to pass in a float array? Or always convert to an array up-front? In our testing we found that converting to a float array upfront was good for performance, since you avoid the unboxing costs of iterating over a List<Float> .

My thinking around this class was if you are using the query vector once, it's the same cost, minus the array allocation, to do the conversion in line.

Users of the class bindings in `ScoreScriptUtils` only pay this cost one time at the cost of the limitation that using different query vectors will [silently fail](https://github.com/elastic/elasticsearch/issues/70437).

Users of this API have to pay that cost on every execution unless we add new painless features.

I could remove this class and add a `toFloatArray()` augmentation to `List` so users will need to call `field('dv').get().dotProduct(params.query.toFloatArray())`.

If we go this route, I'd add those augmentations in a different pr (we'd want ones for `double` array, `long` array and `int` array as well).  What are your thoughts?

jtibshirani commented:
Thanks for the explanations! I agree `field('dv').get().vector[...]` is pretty readable, it works for me. For cosine similarity, I think it is okay to normalize on every call. We've moved towards the messaging of "if you really want to optimize cosine similarity, then normalize all your vectors in advance and use dot product". Cosine similarity is there as a convenience and it's okay if it is not as performant.

Maybe we could catch up offline about QueryVector to discuss the alternatives?

stu-elastic commented:
Based on offline discussion with @jtibshirani,
A) It's preferable to do extra runtime operations to handle Lists than to allocate large arrays for every call for every document.
B) The existing `ScriptScriptUtils` class bindings need to stay approximately as performant as before, even if it costs some code reuse.  Some per-segment overhead is fine.

**A** implies
1) We'll keep `QueryVector` for the new API.
2) The new API can avoid calls into  `VectorUtil` to avoid array allocations

stu-elastic committed:
Rework query vector, add cosineSimilarity, rm iterator, adjust yaml skip, lazy allocate BinaryDenseVector.vector array

stu-elastic committed:
Duplicate float[] and qv methods

stu-elastic committed:
rm statement, test empty cossim

stu-elastic commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/BinaryDenseVector.java:
```diff
@@ -0,0 +1,109 @@
+    public BinaryDenseVector(BytesRef value, int dims, Version indexVersion) {
+        this.value = value;
+        this.indexVersion = indexVersion;
+        this.vector = new float[dims];
```
Done.

stu-elastic commented:
@jdconrad and @jtibshirani I've updated the PR based on the comments above, can you take another look?

* added `cosineSimilarity`, the API can handle pre-normalized query vectors for use from [`ScoreScriptUtils`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-3d28338bcfb4a99c333f39b8f9e15efa817c35cd969f69490cb73694115f4e76R116).  The default behavior is to normalize the query vector while doing the dot product computation.
* there's `QueryVector` and `float[]` versions of [`dotProduct`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR29), [`l1Norm`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR51), [`l2Norm`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR73) and [`cosineSimilarity`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR98).  This avoids forcing users to allocate large float arrays to use those methods as would be necessary if there were only `float[]` versions
* [`QueryVector`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-7ce7eb144f6789254c53bcf950837c57628aedaf8d22580feb175dd3856446e6R17) now is just a thin wrapper around `List`.  The avoids any indirection in the functions above when working with `float[]` avoiding any performance change for existing users of `ScoreScriptUtils`.

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into pain_field_dense_vector

stu-elastic commented:
Do we need [`getVector()`](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR25) at all?  What are people using that for in the old API?

jtibshirani commented:
> Do we need [getVector()](https://github.com/elastic/elasticsearch/pull/83550/files#diff-e6f3cebdf7b7ed4a55c1bf50a4637772a266fb3d853d53a9a683fb9cbf5f2a3cR25) at all? What are people using that for in the old API?

Indeed this is useful to have, people were using it as part of custom calculations (https://github.com/elastic/elasticsearch/issues/51964).

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/BinaryDenseVector.java:
```diff
@@ -0,0 +1,163 @@
+        return dims;
+    }
+
+    protected static ByteBuffer wrap(BytesRef dv) {
```
Does this need to be protected?

stu-elastic replied:
Made `wrap` private.

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/BinaryDenseVector.java:
```diff
@@ -0,0 +1,163 @@
+    /**
+     * dotProduct of doc vector and query vector that normalizes the query vector while performing the calculation.
+     */
+    protected double dotProduct(QueryVector qv, float qvMagnitude) {
```
I don't think we need these special methods that take `qvMagnitude`, because this calculation is the equivalent to just doing a  normal `dotProduct`, then dividing the result by `qvMagnitude`.

stu-elastic replied:
Removed these methods and updated cosSim implementation.

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVector.java:
```diff
@@ -0,0 +1,234 @@
+            checkDimensions(getDims(), qv.size());
+            return dotProduct(qv);
+
+        } else if (queryVector instanceof List<?> list) {
```
Could this just be `instanceof List<Number>`? When we wrote the first version of vectors in painless, I think we could assume the param would have this type.

stu-elastic replied:
We can't due to type erasure.  We can check `instanceof List<?>` and then do an unchecked cast below, adding `@SuppressWarnings("unchecked")`.

If they send in a bad type, they'll get a different reason message.  For the test
```yaml
  - do:
      catch: bad_request
      search:
        body:
          query:
            script_score:
              query: { "exists": { "field": "bdv" } }
              script:
                source: |
                  (int) field(params.field).get().l2Norm(params.query)
                params:
                  query: [1, 2, "three"]
                  field: bdv
```

the error will look like
```
        "failed_shards" : [
          {
            "shard" : 0,
            "index" : "test-index",
            "node" : "ueT4QigUSbK_ngYxjFfLZA",
            "reason" : {
              "type" : "script_exception",
              "reason" : "runtime error",
              "script_stack" : [
                "org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:127)",
                "org.elasticsearch.xpack.vectors.query.DenseVector.l2Norm(DenseVector.java:89)",
                "(int) field(params.field).get().l2Norm(params.query)
",
                "                                             ^---- HERE"
              ],
              "script" : "(int) field(params.field).get().l2Norm(params.query) ...",
              "lang" : "painless",
              "position" : {
                "offset" : 45,
                "start" : 0,
                "end" : 53
              },
              "caused_by" : {
                "type" : "class_cast_exception",
                "reason" : "class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')",
                "stack_trace" : "java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')
	at org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:127)

```

currently, the error looks like
```
        "failed_shards" : [
          {
            "shard" : 0,
            "index" : "test-index",
            "node" : "Fty-bHa-RC6HTekobvpLrw",
            "reason" : {
              "type" : "script_exception",
              "reason" : "runtime error",
              "script_stack" : [
                "org.elasticsearch.xpack.vectors.query.QueryVector.get(QueryVector.java:44)",
                "org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:126)",
                "org.elasticsearch.xpack.vectors.query.DenseVector.l2Norm(DenseVector.java:89)",
                "(int) field(params.field).get().l2Norm(params.query)
",
                "                                             ^---- HERE"
              ],
              "script" : "(int) field(params.field).get().l2Norm(params.query) ...",
              "lang" : "painless",
              "position" : {
                "offset" : 45,
                "start" : 0,
                "end" : 53
              },
              "caused_by" : {
                "type" : "illegal_argument_exception",
                "reason" : "Cannot treat [three] at index [2] of type [java.lang.String] as Number",
                "stack_trace" : "java.lang.IllegalArgumentException: Cannot treat [three] at index [2] of type [java.lang.String] as Number
	at org.elasticsearch.xpack.vectors.query.QueryVector.get(QueryVector.java:44)
```

ie the reason will be
> class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')

vs

> Cannot treat [three] at index [2] of type [java.lang.String] as Number

jtibshirani replied:
Reading quickly, I actually find the first one easier to understand, so happy to go with that if it helps simplify the code.

stu-elastic replied:
Works for me.

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/QueryVector.java:
```diff
@@ -0,0 +1,51 @@
+ * Provide consistent access to query vector elements from a backing list or float array.  Used to provide
+ * a flexible API for DenseVector in painless, which does not have method overloading.
+ */
+public class QueryVector {
```
Is there still a benefit to introducing this wrapper class, or could we just use `List<Number>` directly in the vector functions? (Sorry if I missed this in our conversations)

stu-elastic replied:
The benefit is avoiding an unchecked cast and a nicer error message if the user sends in a mixed type `List`.

We could avoid the wrapper and use a helper instead method, that would still give a nice error and remove the `QueryVector` helper.

stu-elastic replied:
Removed `QueryVector`.

jtibshirani reviewed (commented):
I left some last small comments, apart from these it looks good to me!

stu-elastic committed:
rm inline mag div for dotprod, rm QueryVector, private wrap

jdconrad reviewed (approved):
All the changes look good to me! Thanks for the work on this @stu-elastic

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/KnnDenseVector.java:
```diff
@@ -99,15 +81,15 @@ public double l2Norm(QueryVector queryVector) {
     public double cosineSimilarity(float[] queryVector, boolean normalizeQueryVector) {
         if (normalizeQueryVector) {
-            return dotProduct(queryVector, DenseVector.getMagnitude(queryVector)) / getMagnitude();
+            return (dotProduct(queryVector) / DenseVector.getMagnitude(queryVector)) / getMagnitude();
```
Small comment, we usually would do `dotProduct(queryVector) / (DenseVector.getMagnitude(queryVector) * getMagnitude());` since it's more readable (similar to how you'd write it mathematically) and potentially a little faster.

stu-elastic replied:
Makes sense, changed.

jtibshirani reviewed (approved):
Thanks for all the iterations! We have rally nightly benchmarks for `dense_vector` performance, I will keep an eye on them to double-check there's no performance regression with the original docvalues syntax.

stu-elastic committed:
divide by product in cosine similarity

stu-elastic commented:
@elasticmachine run elasticsearch-ci/part-2
